File: lib/tribucket/mirror.py
"""Mirror provider system with TTL cache and fallback."""
import os
import time
import urllib.request

from tribucket.config import load_json, save_json, mirror_cache_path, mirror_config_path
from tribucket.utils import log


DEFAULT_PROVIDERS = [
    {
        "name": "hunluan",
        "template": "https://gh.do.hunluan.space/https://github.com/{repo}/releases/download/v{version}/{asset}",
        "test_url": "https://gh.do.hunluan.space/",
    },
]


def build_direct_url(repo, version, asset):
    """Build a direct GitHub download URL."""
    return f"https://github.com/{repo}/releases/download/v{version}/{asset}"


def build_mirror_url(template, repo, version, asset):
    """Build a mirror download URL from template."""
    return template.format(repo=repo, version=version, asset=asset)


def test_provider(provider, timeout=3):
    """Test if a mirror provider is reachable. Returns (ok, latency_ms)."""
    test_url = provider.get("test_url", "")
    if not test_url:
        return False, 0

    start = time.monotonic()
    try:
        req = urllib.request.Request(test_url, method="HEAD")
        req.add_header("User-Agent", "tribucket/2.0")
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            latency = int((time.monotonic() - start) * 1000)
            return resp.status < 400, latency
    except Exception:
        latency = int((time.monotonic() - start) * 1000)
        return False, latency


def select_provider(mirror_mode="auto"):
    """Select the best available mirror provider.

    Returns (provider_name, template) or ("direct", None) for direct access.
    """
    if mirror_mode == "direct":
        return "direct", None

    # cn mode: force mirror, skip direct test
    if mirror_mode == "cn":
        user_config = load_json(mirror_config_path(), {})
        providers = user_config.get("providers", DEFAULT_PROVIDERS)
        # Pick first available provider
        for p in providers:
            ok, _ = test_provider(p)
            if ok:
                return p["name"], p.get("template")
        # Fallback to first provider even if not tested
        if providers:
            return providers[0]["name"], providers[0].get("template")
        return "direct", None

    # Load user config
    user_config = load_json(mirror_config_path(), {})
    force = user_config.get("force")

    if force:
        if force == "direct":
            return "direct", None
        providers = user_config.get("providers", DEFAULT_PROVIDERS)
        for p in providers:
            if p["name"] == force:
                return p["name"], p.get("template")
        log(f"Force provider '{force}' not found, falling back to auto")

    # Check cache
    cache = load_json(mirror_cache_path(), {})
    selected = cache.get("selected")
    checked_at = cache.get("checked_at", "")
    ttl = cache.get("ttl_seconds", 3600)

    if selected and checked_at:
        try:
            from datetime import datetime, timezone, timedelta
            dt = datetime.fromisoformat(checked_at)
            if datetime.now(timezone.utc) - dt < timedelta(seconds=ttl):
                log(f"Mirror (cached): {selected}")
                if selected == "direct":
                    return "direct", None
                providers = user_config.get("providers", DEFAULT_PROVIDERS)
                for p in providers:
                    if p["name"] == selected:
                        return p["name"], p.get("template")
        except (ValueError, TypeError):
            pass

    # Probe providers
    providers = user_config.get("providers", DEFAULT_PROVIDERS)
    results = {}
    for p in providers:
        ok, latency = test_provider(p)
        results[p["name"]] = {"ok": ok, "latency_ms": latency}
        log(f"Mirror probe: {p['name']} = {'OK' if ok else 'FAIL'} ({latency}ms)")

    # Also test direct
    direct_ok, direct_latency = _test_direct()
    results["direct"] = {"ok": direct_ok, "latency_ms": direct_latency}
    log(f"Mirror probe: direct = {'OK' if direct_ok else 'FAIL'} ({direct_latency}ms)")

    # Select fastest available
    best_name = None
    best_latency = float("inf")
    for name, result in results.items():
        if result["ok"] and result["latency_ms"] < best_latency:
            best_name = name
            best_latency = result["latency_ms"]

    if not best_name:
        best_name = "direct"

    # Save cache
    _save_mirror_cache(best_name, results)

    if best_name == "direct":
        return "direct", None

    for p in providers:
        if p["name"] == best_name:
            return p["name"], p.get("template")

    return "direct", None


def resolve_download_url(repo, version, asset, mirror_mode="auto"):
    """Resolve the best download URL for an asset.

    Returns (url, provider_name).
    """
    provider_name, template = select_provider(mirror_mode)

    if provider_name == "direct" or not template:
        return build_direct_url(repo, version, asset), "direct"

    url = build_mirror_url(template, repo, version, asset)
    return url, provider_name


def _test_direct(timeout=3):
    """Test direct GitHub access."""
    start = time.monotonic()
    try:
        req = urllib.request.Request("https://github.com", method="HEAD")
        req.add_header("User-Agent", "tribucket/2.0")
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            latency = int((time.monotonic() - start) * 1000)
            return resp.status < 400, latency
    except Exception:
        latency = int((time.monotonic() - start) * 1000)
        return False, latency


def _save_mirror_cache(selected, probe_results):
    """Save mirror probe results to cache."""
    from datetime import datetime, timezone
    cache = {
        "checked_at": datetime.now(timezone.utc).isoformat(),
        "ttl_seconds": 3600,
        "providers": probe_results,
        "selected": selected,
    }
    save_json(mirror_cache_path(), cache)

File: src/engine/mirror.ts
import { loadJson, saveJson } from '../config/store';
import { mirrorCachePath, mirrorConfigPath } from '../config/paths';
import { log, status } from '../utils/log';
import type { MirrorMode } from '../types';

const DEFAULT_PROVIDERS = [
  {
    name: 'hunluan',
    // {tag} is the raw release tag_name (e.g. "v1.2.3", "jq-1.8.1", "15.1.0").
    // We must NOT inject a "v" prefix — release tags are project-specific.
    template: 'https://gh.do.hunluan.space/https://github.com/{repo}/releases/download/{tag}/{asset}',
    test_url: 'https://gh.do.hunluan.space/',
  },
];

/**
 * Build a direct GitHub download URL.
 * `tag` is the raw release tag_name (with whatever prefix the project uses).
 */
export function buildDirectUrl(repo: string, tag: string, asset: string): string {
  return `https://github.com/${repo}/releases/download/${tag}/${asset}`;
}

/**
 * Build a mirror download URL from a provider template.
 * Templates may use {tag} (raw tag_name, preferred) or {version} (legacy,
 * tag with a single leading "v" stripped — kept for backward compat with
 * user mirror.json configs that still reference {version}).
 */
export function buildMirrorUrl(template: string, repo: string, tag: string, asset: string): string {
  const version = tag.replace(/^v/, '');
  return template
    .replace('{repo}', repo)
    .replace('{tag}', tag)
    .replace('{version}', version)
    .replace('{asset}', asset);
}

/**
 * Convert an asset glob pattern (e.g. "fzf-*-windows_amd64.zip") into a RegExp.
 * Only `*` is treated as a wildcard (matches any characters). All other regex
 * metacharacters in the pattern are escaped so they match literally.
 */
function globToRegExp(pattern: string): RegExp {
  const parts = pattern.split('*');
  const escaped = parts.map((p) => p.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'));
  return new RegExp('^' + escaped.join('.*') + '$');
}

/**
 * Resolve an asset_pattern to the actual asset name from a GitHub release.
 *
 * Strategy:
 *  1. Literal exact match: pattern equals an asset name → use it directly.
 *  2. Glob match: pattern contains `*` → find the first asset matching the glob.
 *  3. Suffix match: pattern is a pure suffix (no `/`, no `*`) → find assets
 *     whose name ends with the pattern. Handles the common case where package
 *     definitions list only the platform/arch tail (e.g. "x86_64-pc-windows-msvc.zip")
 *     while the real asset is "bat-v0.26.1-x86_64-pc-windows-msvc.zip".
 *
 * Returns the resolved asset name, or the raw pattern if no release data / no
 * match is available (the caller will then likely 404, but we never invent a
 * fake name).
 */
export function resolveAssetName(releaseData: any | null, pattern: string): string {
  if (!releaseData || pattern === 'NO_MATCH') return pattern;

  const assets: string[] = (releaseData.assets || [])
    .map((a: any) => a.name)
    .filter((n: any): n is string => typeof n === 'string' && n.length > 0);

  if (assets.length === 0) return pattern;

  // 1. Literal exact match.
  if (assets.includes(pattern)) return pattern;

  // 2. Glob match (pattern contains *).
  if (pattern.includes('*')) {
    const re = globToRegExp(pattern);
    const match = assets.find((a) => re.test(a));
    if (match) return match;
  }

  // 3. Suffix match — pattern is a pure suffix with no path separators.
  if (!pattern.includes('/')) {
    const candidates = assets.filter((a) => a.endsWith(pattern));
    if (candidates.length > 0) {
      // Prefer the candidate whose prefix ends in a separator, to avoid
      // e.g. "amd64" matching inside "arm64...amd64".
      const preferred = candidates.find((a) => {
        const prefix = a.slice(0, a.length - pattern.length);
        return prefix === '' || /[-_.]$/.test(prefix);
      });
      return preferred || candidates[0];
    }
  }

  return pattern;
}

async function testProvider(provider: { test_url: string }, timeout = 3000): Promise<[boolean, number]> {
  const start = Date.now();
  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);
    const response = await fetch(provider.test_url, { method: 'HEAD', signal: controller.signal });
    clearTimeout(timeoutId);
    return [response.status < 400, Date.now() - start];
  } catch {
    return [false, Date.now() - start];
  }
}

export async function selectProvider(mirrorMode: MirrorMode = 'auto'): Promise<[string, string | null]> {
  if (mirrorMode === 'direct') return ['direct', null];

  const userConfig = loadJson<any>(mirrorConfigPath(), {});
  const force = userConfig.force;

  if (force === 'direct') return ['direct', null];
  if (force && force !== 'direct') {
    const providers = userConfig.providers || DEFAULT_PROVIDERS;
    const p = providers.find((x: any) => x.name === force);
    if (p) return [p.name, p.template];
  }

  const providers = userConfig.providers || DEFAULT_PROVIDERS;

  // cn mode: force mirror, skip direct, fallback to first provider
  if (mirrorMode === 'cn') {
    for (const p of providers) {
      const [ok] = await testProvider(p);
      if (ok) return [p.name, p.template];
    }
    return providers.length > 0 ? [providers[0].name, providers[0].template] : ['direct', null];
  }

  // auto mode: check cache, then probe
  const cache = loadJson<any>(mirrorCachePath(), null);
  if (cache?.selected) {
    const checkedAt = new Date(cache.checked_at);
    if (Date.now() - checkedAt.getTime() < (cache.ttl_seconds || 3600) * 1000) {
      if (cache.selected === 'direct') return ['direct', null];
      const p = providers.find((x: any) => x.name === cache.selected);
      if (p) return [p.name, p.template];
    }
  }

  // Probe all providers + direct
  status('Testing mirrors...');
  const results: Record<string, { ok: boolean; latency_ms: number }> = {};

  for (const p of providers) {
    const [ok, latency] = await testProvider(p);
    results[p.name] = { ok, latency_ms: latency };
    log(`Mirror probe: ${p.name} = ${ok ? 'OK' : 'FAIL'} (${latency}ms)`);
  }

  const [directOk, directLatency] = await testDirect();
  results['direct'] = { ok: directOk, latency_ms: directLatency };
  log(`Mirror probe: direct = ${directOk ? 'OK' : 'FAIL'} (${directLatency}ms)`);

  // Select fastest
  let bestName = 'direct';
  let bestLatency = Infinity;
  for (const [name, result] of Object.entries(results)) {
    if (result.ok && result.latency_ms < bestLatency) {
      bestName = name;
      bestLatency = result.latency_ms;
    }
  }

  saveJson(mirrorCachePath(), {
    checked_at: new Date().toISOString(),
    ttl_seconds: 3600,
    providers: results,
    selected: bestName,
  });

  if (bestName === 'direct') return ['direct', null];
  const p = providers.find((x: any) => x.name === bestName);
  status(`Mirror selected: ${bestName} (${bestLatency}ms)`);
  return p ? [p.name, p.template] : ['direct', null];
}

async function testDirect(timeout = 3000): Promise<[boolean, number]> {
  const start = Date.now();
  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);
    const response = await fetch('https://github.com', { method: 'HEAD', signal: controller.signal });
    clearTimeout(timeoutId);
    return [response.status < 400, Date.now() - start];
  } catch {
    return [false, Date.now() - start];
  }
}

/**
 * Resolve the final download URL for an asset.
 *
 * @param repo         owner/repo
 * @param tag          raw release tag_name (e.g. "v1.2.3", "jq-1.8.1", "15.1.0")
 * @param pattern      asset_pattern value for the current platform (literal, glob, or suffix)
 * @param mirrorMode   auto / cn / direct
 * @param releaseData  GitHub release object (used to resolve glob/suffix patterns to real asset names)
 */
export async function resolveDownloadUrl(
  repo: string,
  tag: string,
  pattern: string,
  mirrorMode: MirrorMode = 'auto',
  releaseData: any | null = null,
): Promise<[string, string]> {
  const asset = resolveAssetName(releaseData, pattern);
  const [providerName, template] = await selectProvider(mirrorMode);
  if (providerName === 'direct' || !template) {
    return [buildDirectUrl(repo, tag, asset), 'direct'];
  }
  return [buildMirrorUrl(template, repo, tag, asset), providerName];
}
